Fix #2130 and handle NaNs for complex types

mark-poscablo opened the pull request:
Resolves #2130 and also tries to cover the case for complex types as well. Currently, CUDA backend is still being worked on for the complex types.

The essence of the fixes are:
- Guard against the next examined value being a NaN (all backends)
- If the initial min/max value is a NaN (first element of array in CPU, first element that a thread picks up in CUDA and OpenCL) , set it instead to `Binary<T, op>::init()`
- Element index order checks are removed from CUDA backend, since they only make sense to use in a serial reduction (CPU), not in parallel
- The issue only concerns indexed min/max, but OpenCL's non-indexed min/max also had incorrect results also for complex types. Adding the guard fixes it

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -27,6 +27,9 @@ namespace kernel
     static double __host__ __device__ cabs(const cdouble &in) { return (double)abs(in); }
+    template<typename T> __host__ __device__ static bool isNan(T in) { return in != in; }
+    static bool __host__ __device__ isNan(const cfloat &in) { return in.x != in.x || in.y != in.y; }
+    static bool __host__ __device__ isNan(const cdouble &in) { return in.x != in.x || in.y != in.y; }
```
Can you specialize this function so that it is a template? It should look something like this.

```
template<> __host__ __device__ isNan<cfloat>(const cfloat &in) {...}
template<> __host__ __device__ isNan<cdouble>(const cdouble &in) {...}
```

mark-poscablo replied:
Just read up on template specialization. Haha I have a lot to learn in C++, thanks for this tip! I went ahead and did the same too for `cabs()`

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -36,13 +39,17 @@ namespace kernel
         {
+            if (isNan(val)) {
+                Binary<T, op> ireduce;
+                m_val = ireduce.init();
```
I wonder if you could make `init` a static function. That way you don't have to create a binary object. It doesn't really make sense to create a Binary object here.

9prady9 replied:
Yes, it works. I have the changes ready. @mark-poscablo I can send you them as a patch file

mark-poscablo replied:
Good idea - I applied the patch from @9prady9, thanks!

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    //  for complex max op) during the comparisons, and so its 
+    //  magnitude will determine that that element is the max, 
+    //  whereas it should have been ignored since its other 
+    //  part is NaN
```
👍

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    }
+
+    for (int i = 0; i < cols; i++) {
+        ASSERT_FLOAT_EQ(h_min_idx[i], gold_min_idx[i]);
```
idx are integers so you can just use `ASSERT_EQ`

mark-poscablo replied:
Lol I got carried away with using assert floats. Thanks for catching this!

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    }
+
+    for (int i = 0; i < cols; i++) {
+        ASSERT_FLOAT_EQ(h_max_idx[i], gold_max_idx[i]);
```
Not floats so you can just use `ASSERT_EQ`

mark-poscablo replied:
👍

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+TEST(MinMax, MinCplxNaN)
+{
+    float real_wnan_data[] = {
+        849.088836, NAN, 889.189274, NAN, 506.472337,
```
Nit. Is there a significance to the non Nan(non Not a number, just number?) numbers? we can probably just make them simpler because they distract from the actual problem. Not a big deal.

mark-poscablo replied:
Not really, yeah I'll make them simpler as to not distract

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    
+    af::array min_val = af::min(a);
+
+    std::vector< std::complex<float> > h_min_val(cols);
```
you can define the symbols you use using the statements. Like this:
```
using std::vector;
using std::complex;
etc.
```
Then you don't have to specify the namespace.

mark-poscablo replied:
Thanks, this should eliminate some namespace bloat

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    }
+
+    for (int i = 0; i < cols; i++) {
+        ASSERT_FLOAT_EQ(h_min_idx[i], gold_min_idx[i]);
```
Index is an integer so use `ASSERT_EQ`.

mark-poscablo replied:
👍

umar456 commented on test/reduce.cpp:
```diff
@@ -523,6 +523,221 @@ TEST(MinMax, NaN)
+    }
+
+    for (int i = 0; i < cols; i++) {
+        ASSERT_FLOAT_EQ(h_max_idx[i], gold_max_idx[i]);
```
`ASSERT_EQ`

mark-poscablo replied:
👍

mark-poscablo commented:
I'm also going to move the tests for indexed min/max to test/ireduce.cpp btw, since I think it makes more sense to put them there than mix them with regular min/max in test/reduce.cpp

mark-poscablo commented on src/backend/opencl/kernel/morph.hpp:
```diff
@@ -44,7 +44,7 @@ template<typename T, bool isDilation, int SeLength>
 {
     ToNumStr<T> toNumStr;
-    T init = isDilation ? Binary<T, af_max_t>().init() : Binary<T, af_min_t>().init();
+    T init = isDilation ? Binary<T, af_max_t>()::init() : Binary<T, af_min_t>()::init();
```
👍

9prady9 commented:
@umar I think all the changes are in, Can take another look ?

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -7,13 +7,20 @@
 {
-    if ((sabs(lhs[0]) > sabs(rhs)) ||
-        (sabs(lhs[0]) == sabs(rhs) && *lidx < ridx)) {
+    if (!(IS_NAN(lhs[0])) && 
```
I think if we reverse the order of the && operation we can avoid calling IS_NAN as much because of short-circuiting.

so instead of
```
if (!(IS_NAN(lhs[0])) &&
    ((sabs(lhs[0]) < sabs(rhs)) || (sabs(lhs[0]) == sabs(rhs)))) {
```
```
if ((sabs(lhs[0]) < sabs(rhs)) || (sabs(lhs[0]) == sabs(rhs)))
    && !(IS_NAN(lhs[0])))
```

mark-poscablo replied:
Oh yeah that makes sense - that comparison is probably more often useful than the is_nan check anyway. Btw what do you think of making the is_nan check an inline function than a macro? Just to make it safer from potential un-obvious syntactical errors next time we might have to deal with the is_nan check?

umar456 replied:
Fewer macros the better. If you don't run into any issues then you should do it.

umar456 commented on src/backend/opencl/kernel/ireduce_first.cl:
```diff
@@ -52,7 +52,7 @@ void ireduce_first_kernel(__global T *oData,
 
     if (cond && xid < lim) {
-        out_val = iData[xid];
+        if (!(IS_NAN(iData[xid]))) out_val = iData[xid];
```
This condition should wrap both statements in the body of this if statement. Probably a good idea to add a test for this scenario.

mark-poscablo replied:
So this last one - let me try to see if I got what you're trying to say here. Do you mean that both assignments to `out_val` and `out_idx` should be together in the body of an if-statement?

mark-poscablo replied:
:+1: changed this based on our slack conversation. The is_nan check be moved to the outer condition so that we don't assign out_val and out_idx to new values if the current iData element is nan. just waiting for the checks now

umar456 commented on test/binary_ops.hpp:
```diff
@@ -85,7 +85,7 @@ struct Binary<T, AF_BINARY_MAX>
     struct Binary<T, AF_BINARY_MIN>                 \
     {                                               \
-        T init()                                    \
+        static T init()                                    \
```
The backslash is not formatted correctly.

mark-poscablo replied:
Yup I'll fix these slashes

mark-poscablo replied:
👍

umar456 commented on test/ireduce.cpp:
```diff
@@ -14,7 +14,16 @@
+using af::constant;
+using af::span;
+using af::min;
+using af::allTrue;
```
:+1:

umar456 commented on test/ireduce.cpp:
```diff
@@ -185,3 +194,146 @@ TEST(IndexedReduce, MaxReduceDimensionHasSingleValue)
+    int cols = 2; 
+    array a(rows, cols, test_data); 
+ 
+    float gold_max_val[] = { 5.0, 0.0 }; 
```
add f to the end of those constants to avoid narrowing warnings. i.e. `{5.0f, 0.f};`

umar456 replied:
Seems like there are other places where this could be applied.

mark-poscablo replied:
👍

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -57,13 +79,16 @@ namespace kernel
         {
-            if (cabs(val) > cabs(m_val) ||
-                (cabs(val) == cabs(m_val) &&
-                 idx <= m_idx)) {
```
Why was this last condition removed? I think it enforces that if the values are equal then we should prefer indices that are smaller.

umar456 replied:
Another good candidate for a test.

mark-poscablo replied:
I remember that the CUDA reduce and ireduce tests were failing because of that last condition right after I added the isNan condition and the corresponding initial test cases. In any case, this works in a serial reduction, but in CUDA (and I think in a parallel setting in general), that last condition gives incorrect output in the following scenario: suppose that `val > m_val` but `idx > m_idx` - this case is very plausible since we're not going through the array sequentially. In this case, we should assign `m_val` and `m_idx` to `val` and `idx`, but that last condition will prevent that from happening since `idx > m_idx`. The min operation suffered a similar problem. Also I think this is the reason why the OpenCL min/max ireduce kernels don't have this last condition (even before I modified them). Please let me know your thoughts?

umar456 replied:
The (idx) only comes into play when the `val` and `m_val` values are the same (notice that the right-hand side is surrounded by the parenthesis). The new condition should look something like this:
```
(cabs(val) > cabs(m_val) || (cabs(val) == cabs(m_val) && idx <= m_idx)) && !isNan(val)
```

mark-poscablo replied:
OH that makes so much more sense. I think that confused me at first and tried to change the parenthesizing of the conditions, which led me to the failing tests I mentioned earlier. I'll the parenthesizing back (as you've written above) and I'll add the conditions back too. Thanks for the clarification!

mark-poscablo replied:
Actually I'm wrong on another note - the opencl kernel had that last condition also. Will add that back there too.

mark-poscablo replied:
And yep I can add tests for this specific scenario

mark-poscablo replied:
Looking more closely at the original version of the kernels before i modified them, looks like the kernels actually prefer the larger index (the newer element) in case `val == `m_val`. I'll go with that original intent unless you want me to change it to really prefer smaller indices instead?

mark-poscablo replied:
Actually, hold on - that above comment is true for min, but the max is the other way - it prefers the smaller index if available.

mark-poscablo replied:
:+1: just waiting for these changes to pass checks now

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -36,13 +55,16 @@ namespace kernel
         {
-            if (cabs(val) < cabs(m_val) ||
-                (cabs(val) == cabs(m_val) &&
-                 idx > m_idx)) {
```
I think this condition was incorrectly removed. See comment below.

mark-poscablo replied:
Addressed this concern in my comment in the related change too (the max counterpart), please let me know what you think in that thread.

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -22,11 +22,30 @@ namespace cuda
 namespace kernel
 {
+    template<typename T> __host__ __device__ 
+    static double cabs(const T& in) { return (double)in; }
```
@pavanky I am questioning the reasoning behind using doubles as the output of this function. Integral types should return the same type. I understand wanted to keep the original C return value but we are leaving a lot of performance on the table here. @mark-poscablo I am making general comments. You don't need to change anything in this PR.

umar456 commented on src/backend/cpu/kernel/ireduce.hpp:
```diff
@@ -49,13 +55,17 @@ struct MinMaxOp<af_max_t, T>
         m_val(val), m_idx(idx)
     {
+        if (isNan(val)) {
+            m_val = Binary<T, af_max_t>::init();
```
The m_idx value should be updated as well. To what is a good question.

umar456 replied:
@pavanky perhaps max uint?

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -57,13 +79,16 @@ namespace kernel
             m_val(val), m_idx(idx)
         {
+            if (isNan(val)) {
+                m_val = Binary<T, af_max_t>::init();
```
Update the value of m_idx

umar456 commented on src/backend/cpu/kernel/ireduce.hpp:
```diff
@@ -28,13 +30,17 @@ struct MinMaxOp
-        if (cabs(val) < cabs(m_val) ||
-            (cabs(val) == cabs(m_val) &&
-             idx > m_idx)) {
+        if (!isNan(val) &&
```
I think if we reverse the order of the && operation we can avoid calling IS_NAN as much because of short-circuiting.

so instead of
```
if (!(IS_NAN(lhs[0])) &&
    ((sabs(lhs[0]) < sabs(rhs)) || (sabs(lhs[0]) == sabs(rhs)))) {
```
```
if ((sabs(lhs[0]) < sabs(rhs)) || (sabs(lhs[0]) == sabs(rhs)))
    && !(IS_NAN(lhs[0])))
```

umar456 commented on src/backend/cpu/kernel/ireduce.hpp:
```diff
@@ -49,13 +55,17 @@ struct MinMaxOp<af_max_t, T>
-        if (cabs(val) > cabs(m_val) ||
-            (cabs(val) == cabs(m_val) &&
-             idx <= m_idx)) {
+        if (!isNan(val) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -36,13 +55,16 @@ namespace kernel
-            if (cabs(val) < cabs(m_val) ||
-                (cabs(val) == cabs(m_val) &&
-                 idx > m_idx)) {
+            if (!isNan(val) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

umar456 commented on src/backend/cuda/kernel/ireduce.hpp:
```diff
@@ -57,13 +79,16 @@ namespace kernel
-            if (cabs(val) > cabs(m_val) ||
-                (cabs(val) == cabs(m_val) &&
-                 idx <= m_idx)) {
+            if (!isNan(val) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -23,8 +30,9 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
 {
-    if ((sabs(lhs[0]) < sabs(rhs)) ||
-        (sabs(lhs[0]) == sabs(rhs) && *lidx > ridx)) {
+    if (!(IS_NAN(lhs[0])) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -35,8 +43,9 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
 {
-    if ((*lhs > rhs) ||
-        (*lhs == rhs && *lidx < ridx)) {
+    if (!(IS_NAN(*lhs)) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -46,8 +55,9 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
 {
-    if ((*lhs < rhs) ||
-        (*lhs == rhs && *lidx > ridx)) {
+    if (!(IS_NAN(*lhs)) &&
```
Switch order of isNan operation. See other comments for info.

mark-poscablo replied:
👍

mark-poscablo commented:
Just rebased this PR off the current master. Build went fine on my OS X and will push it now to my fork so it can run against the checks again.

umar456 commented on test/scan_by_key.cpp:
```diff
@@ -106,7 +106,7 @@ void verify(af::dim4 dims,
                     ASSERT_NEAR(gold, out[index], eps);
                 } else {
-                    gold = binOp.init();
+                    gold = Binary<To, op>::init();
```
Binary is not a public facing API and should not be used in test. Please remove it.

mark-poscablo replied:
So I think this test is using a separate header file that defines its own templated Binary operators (test/binary_ops.hpp), apart from the ArrayFire source. I accidentally made this static too in the course of scanning the whole repo for all Binary operator implementations. Let me just revert this and test/binary_ops.hpp to it's original version instead then?

mark-poscablo commented:
@umar456 All checks pass, please review again whenever you're free and let me know what you think - feel free to keep them coming until we get quality code for this PR

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -7,13 +7,21 @@
-    if ((sabs(lhs[0]) > sabs(rhs)) ||
-        (sabs(lhs[0]) == sabs(rhs) && *lidx < ridx)) {
+    if (((sabs(lhs[0]) > sabs(rhs)) ||
+         (sabs(lhs[0]) == sabs(rhs) &&
```
Nit: I liked it better when the lidx < ridx condition was on the same line as the equality condition.

mark-poscablo replied:
:+1: Yeah that makes sense. I think I was just following suit with the other backends on how this condition was originally written. I could change the others too to this style.

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -23,20 +31,22 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
-    if ((sabs(lhs[0]) < sabs(rhs)) ||
-        (sabs(lhs[0]) == sabs(rhs) && *lidx > ridx)) {
+    if (((sabs(lhs[0]) < sabs(rhs)) ||
+         (sabs(lhs[0]) == sabs(rhs) &&
```
Nit: I liked it better when the condition was on the same line.

mark-poscablo replied:
:+1:

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -23,20 +31,22 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
-        (*lhs == rhs && *lidx < ridx)) {
+    if (((*lhs > rhs) ||
+         (*lhs == rhs && *lidx < ridx)) &&
+        !is_nan(*lhs)) {
```
Is rhs guaranteed to be a non-nan number? Do we need to add a check for that as well?

mark-poscablo replied:
Actually... the way this was originally written kinda made me confused because the order of lhs and rhs is flipped on the other backends. In any case though, I think I should be checking if `rhs` is nan here, instead of `*lhs`. `*lhs` (pointing to the kernel's `out_val`) is actually guaranteed to be non-nan, since it's initialized to the type's reduction initial value in the beginning (in `ireduce_first_kernel()`), and there is a guard there that prevents it from being changed to an element of the input array if that element is nan. The same guard then should be applied in the operator (in this case checking if `rhs` is nan) to maintain that guarantee.

umar456 commented on src/backend/opencl/kernel/iops.cl:
```diff
@@ -46,8 +56,9 @@ void binOp(T *lhs, uint *lidx, T rhs, uint ridx)
-        (*lhs == rhs && *lidx > ridx)) {
+    if (((*lhs < rhs) ||
+         (*lhs == rhs && *lidx > ridx)) &&
+        !is_nan(*lhs)) {
```
Is rhs guaranteed to be a non-nan number? Do we need to add a check for that as well?

mark-poscablo commented:
@umar456 I've addressed your latest comments and I just rebased my changes onto the latest master. All checks pass, please review again whenever you have time. Thanks!

umar456 commented on test/reduce.cpp:
```diff
@@ -508,7 +509,7 @@ TYPED_TEST(Reduce, Test_Any_Global)
 }
 
-TEST(MinMax, NaN)
+TEST(MinMax, MinMaxNaN)
```
Nit: this doesn't really add information to the test name. I am ok with it going in.
